Weapon stats

peter-budo opened the pull request:
Weapon statistics available. 

Summary:
- image resources
- created BaseIntelAdapter to avoid hustle of these inherited methods
- weapon fragment and adapter added
- made loading progress layout on its own so can be added to any layout where need it through <include>

peter-budo committed:
Image resources for weapon stats

peter-budo committed:
Temporary commit before switching machines

peter-budo committed:
Resized weapon images canvas from 256x256 to 256x140 as there was too much empty space around them. Which made it rather hard to put them to reasonable list item size

peter-budo committed:
- Added BaseIntelAdapter to reduce noise of override methods, already existing adapter should be corrected in pull request after this one
- BaseListFragment provides now loaders functionality
- Made Weapon class comparable when there is same kills count it will compare based on weapon time equipped (this is to make list of weapon as close match as possible, an extra functionality not in the OTHER dude app)
- Added ui classes and layouts of weapon statistics

peter-budo committed:
- Hide ProgressBar when weapon has at least one service star
- Removed unlock progress I think it is not need it, and let progress bar to match parent

peter-budo committed:
Fixed vertical alignment of service stars count box

peter-budo committed:
Merge branch 'dev' into feature_weapon_stats

Conflicts:
	BF4Intel/src/main/java/com/ninetwozero/bf4intel/factories/FragmentFactory.java
	BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/fragments/NavigationDrawerFragment.java
	BF4Intel/src/main/res/values/content_strings.xml

peter-budo committed:
This class wasn't committed with rest of merge (no idea why)

peter-budo committed:
Merge branch 'dev' into feature_weapon_stats

peter-budo committed:
- Moved WeaponsImageMap to rest of the resources, replace direct call for map with getter to avoid missing resource crash and dropped temporary map
- Added loading indicator to weapons list
- Extracted loading indicator to layout on its own and updated files that been using this code with <include> version

karllindmark commented on core/src/test/java/com/ninetwozero/bf4intel/json/weaponstats/WeaponStatisticsTest.java:
```diff
@@ -0,0 +1,23 @@
+    @Test
+    public void should_parse_json() throws IOException {
+        WeaponStatistics stats = IntelJsonParser.parse("/json/weapon-stats.json", WeaponStatistics.class);
+        assertEquals(WEAPON_COUNT, stats.getWeaponsList().size());
```
:+1:

karllindmark commented on core/src/main/java/com/ninetwozero/bf4intel/json/weaponstats/Weapon.java:
```diff
@@ -0,0 +1,63 @@
+
+import com.google.gson.annotations.SerializedName;
+
+public class Weapon implements Comparable<Weapon> {
```
Neat!

karllindmark commented on BF4Intel/src/main/res/layout/fragment_list_stats.xml:
```diff
@@ -0,0 +1,21 @@
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+  android:layout_width="match_parent"
+  android:layout_height="match_parent"
+  android:background="@drawable/test_soldier_bg"
```
He doesn't really fit in :(

peter-budo replied:
Removed

karllindmark commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/stats/weapons/WeaponStatsFragment.java:
```diff
@@ -0,0 +1,89 @@
+        showLoadingStateInActionBar(false);
+        List<Weapon> weaponList = ws.getWeaponsList();
+        Collections.sort(weaponList);
+        adapter = new WeaponStatsAdapter(weaponList, getContext());
```
Do we really want to create a new adapter each time we've downloaded some data? I think we could create one during `onCreateView()`, and then just do some `((WeaponStatsAdapter)getListAdapter()).setItems(weaponList);` in the `processResult()`.

peter-budo replied:
You may cause a crash if context does change, so it is safer in my opinion to created new adapter

karllindmark replied:
:+1:

karllindmark commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/resources/maps/weapons/WeaponsImageMap.java:
```diff
@@ -0,0 +1,154 @@
+    };
+
+    public static int get(final String key) {
+        return map.containsKey(key) ? map.get(key) : R.drawable.acc_none;
```
:+1:

karllindmark commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/stats/weapons/WeaponStatsFragment.java:
```diff
@@ -0,0 +1,89 @@
+import java.util.Collections;
+import java.util.List;
+
+public class WeaponStatsFragment extends BaseListFragment {
```
Take a quick look at `BaseLoadingListFragment` and see if you can use it here. :-)

karllindmark replied:
I think it'll add the rotating icon in the `ActionBar` as well.

peter-budo replied:
I say blame the fool whom created `BaseListFragment` and `BaseLoadingListFragment`

karllindmark replied:
AP: Create proposal of further usage.

karllindmark commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/stats/weapons/WeaponStatsAdapter.java:
```diff
@@ -0,0 +1,50 @@
+        }
+        Weapon weapon = itemsList.get(position);
+
+        ((TextView) view.findViewById(R.id.index)).setText(String.valueOf(position + 1));
```
Supposed to be a new line between each row? L31-39

peter-budo replied:
I can remove this if you want. It is just "sanity" concept that I have when working with UI seems to make it easier to comprehend what is belonging to what

karllindmark commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/stats/weapons/WeaponStatsAdapter.java:
```diff
@@ -0,0 +1,50 @@
+
+        ((TextView)view.findViewById(R.id.weapon_kills)).setText(String.valueOf(weapon.getKills()));
+
+        if(weapon.getServiceStarsCount() > 0) {
```
Do we really want to hide the `ProgressBar` when we get beyond our first service star? I found it kinda confusing when I saw that some weapons lower down had it. :cat:

peter-budo replied:
Agree

karllindmark commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/stats/SoldierStatisticsActivity.java:
```diff
@@ -84,13 +78,14 @@ private void setupActionBar() {
                     new ActionBar.TabListener() {
                         @Override
-                        public void onTabSelected(final ActionBar.Tab tab, final android.app.FragmentTransaction ft) { final FragmentManager manager = getSupportFragmentManager();
+                        public void onTabSelected(final ActionBar.Tab tab, final android.app.FragmentTransaction ft) {
```
I'm a bit worried that we're using the wrong `FragmentTransaction` here - will that cause any problems? I think there's supposed to be one under `support.v4.app`. :question:

peter-budo replied:
I just wonder who was that dude that wrote that mambo-jambo piece of code...

karllindmark replied:
That must've been from before when we weren't using the support.v4.app.*Fragment:s. Feel free to change!

karllindmark replied:
OK, we'll wait with this until we have the ViewPager in place.

karllindmark commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/battlefeed/layouts/BattlePackLayout.java:
```diff
@@ -86,7 +86,7 @@ private void populateDataTable(final Context context, final View view, final Lis
                     );
                 } else if(category.equals("weapon")) {
-                    populateCell(cell, WeaponImageMap.get(itemKey), WeaponStringMap.get(itemKey), NO_SUBTITLE);
+                    populateCell(cell, WeaponsImageMap.get(itemKey), WeaponStringMap.get(itemKey), NO_SUBTITLE);
```
:+1:

karllindmark commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/base/adapter/BaseIntelAdapter.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @Override
+    public Object getItem(int position) {
```
This should return T.

peter-budo replied:
Done

karllindmark commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/base/adapter/BaseIntelAdapter.java:
```diff
@@ -0,0 +1,39 @@
+
+import java.util.List;
+
+public class BaseIntelAdapter<T> extends BaseAdapter {
```
Make this abstract?

peter-budo replied:
Why?

karllindmark replied:
We are actually using it as a base class - might as well give it the abstract keyword as well?

karllindmark commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/base/adapter/BaseIntelAdapter.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @Override
+    public long getItemId(int position) {
```
Provide abstract template for subclasses to implement?

peter-budo replied:
Why?

karllindmark replied:
The defaults aren't really usable, so it's better to _force_ implementation further down the road.

karllindmark commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/base/adapter/BaseIntelAdapter.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @Override
+    public View getView(int position, View convertView, ViewGroup parent) {
```
Provide abstract template for subclasses to implement?

karllindmark commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/base/ui/BaseListFragment.java:
```diff
@@ -2,17 +2,27 @@
+import com.ninetwozero.bf4intel.R;
+import com.ninetwozero.bf4intel.utils.Result;
+
+public abstract class BaseListFragment extends ListFragment implements LoaderManager.LoaderCallbacks<Result> {
```
Check out `BaseListLoadingFragment`. This class was intended for static data (that didn't need any "loading" capabilities). `getContext()` would be a good idea to bring with you. :-)

peter-budo replied:
Then original attempt was fail because list is expected to load static data in most of the cases. Given the aim of our app that mean 99%

karllindmark commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/factories/FragmentFactory.java:
```diff
@@ -65,10 +68,8 @@ public static Fragment get(final Type type, final Bundle data) {
-            //    return SoldierOverviewFragment.newInstance(data);
-            /*case SOLDIER_ASSIGNMENTS:
-                return AssignmentsFragment.newInstance(data);*/
+            case WEAPON_STATS:
```
Don't forget to pass data!

peter-budo replied:
This was a trap ready for you and you fall for it :ghost: WeaponsStatsFragment by example of other Fragment created by @karllindmark has inside this hidden gem https://github.com/karllindmark/com.ninetwozero.battlelog/blob/ae0f23b6aef9c944a73b72cf3134224101879529/BF4Intel/src/main/java/com/ninetwozero/bf4intel/ui/stats/weapons/WeaponStatsFragment.java#L31

Which mean I do not have to provide new Bundle if I do not want. Nevertheless we should consolidate "single avenue" for future

karllindmark commented:
##### It seems to be sturdy!

No crashes. :cake: 

---

A few questions (and thoughts - I'm thinking out loud a bit so that we can bounce ideas):

##### BF4Intel/src/main/res/drawable/servicestar.png

There already is a service star in the drawable directory (higher resolution too) - award_service_star.png. We could rename it just servicestar.png and use it through-out the app.

##### ViewPager for swipe:able tabs?

Since we're using the support library nowadays (or more so - the support Fragments), we can also use the ViewPager with them. Could you add it to the layout, so that we can both swipe and select a tab to browse the stats? 

It was really bugging me when I was mocking the layouts that I couldn't use the `ViewPager` (due to regular `app.*Fragment` and not `support.v4.app.*Fragment`), so I think it's a great gain to include it now (with `FragmentPagerAdapter`). Also, the `myViewPager.setOffscreenPageLimit(2)` is great to ensure that we don't re-download data in the weapons tab just because we scrolled to the reports tab. The 2 means that it will save up to 5 Fragments, as seen below:

```
[s][s][v][s][s]
[ ][ ][s][s][v]
```

##### Handle bundle with data via getArguments()

We should probably implement this functionality at this point already, so that we don't need to go back just for this later. See L78-100 in `SoldierOverviewFragment` for more information on how it's solved in other parts of the app. :-)

##### Increase layout (read: text) readability

It looks like the text is a bit too light - I tried using `@color/darkgrey` and it seemed to be a bit more readable. Could you try it out, and let me know how it looks on your end?

### Suitable as "card"?

I'm not quite sure, but maybe it is. Here's the layout code needed for it to work:

`fragment_list_stats.xml`:

```
<?xml version="1.0" encoding="utf-8"?>
<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
  android:layout_width="match_parent"
  android:layout_height="match_parent"
  android:orientation="vertical"
  android:padding="@dimen/spacing_m"
  android:background="@color/lightgrey">

  <ListView
    android:id="@android:id/list"
    android:layout_width="match_parent"
    android:layout_height="wrap_content"
    android:divider="@color/lightgrey"
    android:dividerHeight="@dimen/spacing_m"/>

  <TextView
    android:id="@android:id/empty"
    android:layout_width="0dp"
    android:layout_height="0dp" />

  <include layout="@layout/loading_progress" />

</LinearLayout>
```

Plus, the root `LinearLayout` in `list_item_weapon.xml` would need the following attribute:

```
style="@style/WhiteContainer"
```

Not quite sure what to think of it as "card". Reports are suitable as cards (as seen in feed), and this _might_ be too.

peter-budo committed:
- Base adapter instead of generic object returns actual item of the list
- ProgressBar now visible for all items
- Changed colour of text on list item, tweaked colour of service star box, moved kill count more of the right side border
- Renamed award_service_star and deleted servicestar

peter-budo committed:
Applied card board layout

peter-budo committed:
- Moved WeaponStatsFragment to BaseLoadingListFragment
- Removed obsolete methods from BaseListFragment

peter-budo committed:
Abstracted BaseIntelAdapter

peter-budo committed:
Removed empty lines

peter-budo committed:
Removed invocation of fragment from FragmentFactory without bundle

peter-budo committed:
Temporary commit while moving between computers

peter-budo committed:
Assignments URI

peter-budo committed:
Assignments and awards URI provided

peter-budo committed:
Updated UrlFactory and all dependent classes

Commit:
Make getItemId(int) abstract instead of getItem(int)
